consul/prepared_query_endpoint.go

```diff
@@ -0,0 +1,581 @@
+package consul
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/armon/go-metrics"
+	"github.com/hashicorp/consul/consul/structs"
+)
+
+var (
+	// ErrQueryNotFound is returned if the query lookup failed.
+	ErrQueryNotFound = errors.New("Query not found")
+)
+
+// PreparedQuery manages the prepared query endpoint.
+type PreparedQuery struct {
+	srv *Server
+}
+
+// Apply is used to apply a modifying request to the data store. This should
+// only be used for operations that modify the data. The ID of the session is
+// returned in the reply.
+func (p *PreparedQuery) Apply(args *structs.PreparedQueryRequest, reply *string) (err error) {
+	if done, err := p.srv.forward("PreparedQuery.Apply", args, args, reply); done {
+		return err
+	}
+	defer metrics.MeasureSince([]string{"consul", "prepared-query", "apply"}, time.Now())
+
+	// Validate the ID. We must create new IDs before applying to the Raft
+	// log since it's not deterministic.
+	if args.Op == structs.PreparedQueryCreate {
+		if args.Query.ID != "" {
+			return fmt.Errorf("ID must be empty when creating a new prepared query")
+		}
+
+		// We are relying on the fact that UUIDs are random and unlikely
+		// to collide since this isn't inside a write transaction.
+		state := p.srv.fsm.State()
+		for {
+			args.Query.ID = generateUUID()
+			_, query, err := state.PreparedQueryGet(args.Query.ID)
+			if err != nil {
+				return fmt.Errorf("Prepared query lookup failed: %v", err)
+			}
+			if query == nil {
+				break
+			}
+		}
+	}
+	*reply = args.Query.ID
+
+	// Grab the ACL because we need it in several places below.
+	acl, err := p.srv.resolveToken(args.Token)
+	if err != nil {
+		return err
+	}
+
+	// Enforce that any modify operation has the same token used when the
+	// query was created, or a management token with sufficient rights.
+	if args.Op != structs.PreparedQueryCreate {
+		state := p.srv.fsm.State()
+		_, query, err := state.PreparedQueryGet(args.Query.ID)
+		if err != nil {
+			return fmt.Errorf("Prepared Query lookup failed: %v", err)
+		}
+		if query == nil {
+			return fmt.Errorf("Cannot modify non-existent prepared query: '%s'", args.Query.ID)
+		}
+		if (query.Token != args.Token) && (acl != nil && !acl.QueryModify()) {
```

Comment:
Also if the above is true, and we basically apply a blanket-write policy to the ACL's, then not sure it makes sense to check if the token is the original creator since it could have been revoked by this point.